### ISSUE TITLE
Variables

### ISSUE THREAD

**Issue by daveyarwood** [alda-lang/alda#131]
Original idea (originally proposed by me in #7, now being broken out into its own issue):

> **Variables**

> Variables should work like they do in Lilypond. Any piece of code can be enclosed in brackets and defined as a variable, then called in the code any number of times, at any point in time.

> Desired syntax:

> ```
# Defining a variable: 
riffA = [f8 f g+ a > c c d c <]
riffB = [b-8 b- > c+ d f f g f <]
riffC = [> c8 c d+ e g g a g < ]
riffD = [f8 f g+ a > c c d < b > | c c < b- b- a a g g]

> # calling a variable
$riffA # expands to "f8 f g+ a > c c d c <"

> # using variables + repeats
($riffA)x4 
($riffB)x2  ($riffA)x2
$riffC $riffB $riffD

> # bonus: conglomerate variables into other variables!
rockinRiff = 
  [ ($riffA)x4 
    ($riffB)x2  ($riffA)x2
    $riffC $riffB $riffD    ] 

> guitar/saxophone: 
    ($rockinRiff) x9999999
```

That was from January, and since then Alda's syntax has evolved a little. Parentheses are no longer available in the way proposed above (parentheses now mean inline Clojure code).

New proposed syntax, incorporating ideas proposed by @MillerMark in #7:

```
# Defining a variable: 
riffA = f8 f g+ a > c c d c <
riffB = b-8 b- > c+ d f f g f <
riffC = > c8 c d+ e g g a g <
riffD = f8 f g+ a > c c d < b > | c c < b- b- a a g g

# calling a variable
riffA # expands to "f8 f g+ a > c c d c <"

# using variables + repeats
riffA*4
riffB*2 riffA*2   # or riffB** riffA**
riffC riffB riffD

# bonus: conglomerate variables into other variables!
rockinRiff = [ 
    riffA*4
    riffB*2 riffA*2
    riffC riffB riffD
] 

guitar/saxophone: 
    rockinRiff*9999999
```

Key points:
- `$` is not needed before variable names -- whether a variable is being defined or evaluated can be left to context
- brackets around variable value can be omitted when it's all on one line -- cleaner than always requiring them
- `*2` syntax instead of `x2` -- easier to differentiate visually, and makes it possible to omit the space between "thing-being-repeated" and "times-N"

I'm totally sold on this new syntax. It cleaner-looking, easier to read, and quicker to type. A little added complexity in the parser is well worth the simpler syntax that it gives the user.

**Comment by daveyarwood** [comment]
To pave the way for transposition operators (#129), @MillerMark proposes that hyphens not be allowed in variable names -- at this point, the idea is that variable names must start with two letters, and may contain letters, digits and underscores.

**Comment by elyisgreat** [comment]
Why not have transposition be an attribute? To deal with individual instruments, you could have instrument-wide attributes that weren't global.

**Comment by daveyarwood** [comment]
@elyisgreat I like that idea. I think we should implement it both ways (as an attribute, and as an add-on for blocks of notes), for greater flexibility/expressibility. By the way, attributes can be applied both per-instrument and globally.

**Comment by daveyarwood** [comment]
Started working on implementing variables today, which led me down the rabbit hole of refactoring and improving Alda's parser. It seems that the parser, in its current state, is not equipped to handle only reading events on one line (i.e. `variableName = every event through the end of this line`), which is something that would be really nice to have for variables. 

Fumbling around with this led me to the thought that maybe our language has become too big for a single parser, and it would probably be a good idea for us to break parsing out into a bunch of discrete steps, rather than doing it all in one go with a single, monolithic grammar. I think this would make the grammar easier to reason about and modify, and should improve performance as well; we could even leverage asynchronous parsing of different parts of a score for an added perf boost. Currently, it takes me about 6 seconds (on my 2014 Macbook Pro) to parse `test/examples/bach_cello_suite_no_1.alda`, which is not great, IMO.

Per [Instaparse's Performance Tip No. 10](https://github.com/Engelberg/instaparse/blob/master/docs/Performance.md#performance-tips), feeding an Instaparser smaller chunks of text ought to improve performance.

I've already spent hours on this today -- turns out that rebuilding the way our parser works from the bottom up is *super hard* -- but I'll continue to tinker with it. I think it will be well worth the effort.

**Comment by daveyarwood** [comment]
The parser changes I mentioned above have been implemented -- I should be getting to this feature soon.

**Comment by daveyarwood** [comment]
Moving this back to "up for grabs" for now as I work on #153. I have some work on variables on a feature branch, and plan to resume working on it after I'm done with the stuff I'm working on at the moment.

**Comment by elyisgreat** [comment]
By the way, I think we'll need a new "natural" syntax, as I think it makes more sense for `=` to be used exclusively for declarations. `º` is a good syntax, however I don't know how easily typeable it is. perhaps `0`, as it can't be used for durations, but I think it's confusing. Any thoughts?

**Comment by daveyarwood** [comment]
I think non-standard Unicode characters are out of the question, for reasons of usability. I'm inclined to agree that we should use a different character to represent naturals once we're also using `=` for variables -- also using it to mean "natural" will definitely lead to confusion and mistakes. I also agree 0 would be too confusing. I'll have to think about this... Off the top of my head, backtick might be an option: ```c` ```  Or perhaps apostrophe? `c'` I don't think we're using either of those.

**Comment by elyisgreat** [comment]
You mean ASCII? because `º` is U+00BA in Unicode. I don't know what you mean by "non-standard". 

I think the `\`` is the best bet. `'` could be interpreted as a string separator (for example if someone wanted to use `piano 'piano1':` instead of `piano "piano1":`)

**Comment by daveyarwood** [comment]
By "non-standard" here I just mean characters where it's not immediately obvious how to type them :)

I'm leaning toward ``` ` ``` here, unless a better idea surfaces. We have plenty of time to think about this one -- I think it will be a while before support lands for variables.

**Comment by daveyarwood** [comment]
I think I will be getting back to working on this feature soon (finally). @elyisgreat, what do you think about using `_` to mean natural?

```
(key-sig "f+ c+ g+")
a b > c_ d e f_ g_ a
```

I think I like the way that looks the best.

`_` can also be used as part of a variable name, but the parser is set up in a way that it won't be ambiguous, so it shouldn't be a problem.

**Comment by elyisgreat** [comment]
Was going to mention how I always think of _ as a space character for variable names, so it looks kind of weird to me.

**Comment by Pomax** [comment]
as a small outsider comment, the backtick is a horrible symbol exactly because it conflicts with quotation marks so much. If the choice is between "variables need to start with two letters" or "variables can be anything that starts with `$` until the next white space characters", the latter is infinitely preferable.

**Comment by daveyarwood** [comment]
Variables are now available in [1.0.0-rc21](https://github.com/alda-lang/alda/releases/tag/1.0.0-rc21)!

Interestingly, I didn't end up needing to change the syntax for naturals. The equal sign `=` is now doing double-duty as the variable definition operator and the natural sign, depending on whether it comes after a variable name or a note.

e.g.

```
# variable definition
myVariable = c d e f g

# natural sign
(key-sig [:b :flat :minor])
b > c d e f g a= b

# double duty
myVariable = b > c d e f g a= b
```

Because variable names are required to start with at least two letters, there is no conflict in the grammar.

Should we still change the natural sign, to avoid confusion between these 2 uses of `=`? I think probably yes, and I would still lean towards underscore. So far, I don't particularly mind the double meaning of `=`, but I'm definitely open to changing that in the near future, if people feel strongly about it. We can discuss in a separate issue, whoever is interested :)

**Comment by elyisgreat** [comment]
now that variables are implemented, what is to be the "natural" syntax?
